diff -up util-linux-ng-2.17.2/mount/mount.8.kzak util-linux-ng-2.17.2/mount/mount.8
--- util-linux-ng-2.17.2/mount/mount.8.kzak	2010-03-22 09:17:23.000000000 +0100
+++ util-linux-ng-2.17.2/mount/mount.8	2010-12-02 15:54:44.000000000 +0100
@@ -941,6 +941,9 @@ options from command line (
 .B ro
 Mount the filesystem read-only.
 .TP
+.B _rnetdev
+Like _netdev, except "fsck -a" checks this filesystem during rc.sysinit.
+.TP
 .B rw
 Mount the filesystem read-write.
 .TP
diff -up util-linux-ng-2.17.2/mount/mount.c.kzak util-linux-ng-2.17.2/mount/mount.c
--- util-linux-ng-2.17.2/mount/mount.c.kzak	2010-03-22 09:10:56.000000000 +0100
+++ util-linux-ng-2.17.2/mount/mount.c	2010-12-02 15:54:44.000000000 +0100
@@ -154,6 +154,8 @@ static const struct opt_map opt_map[] = 
   { "group",	0, 0, MS_GROUP  },	/* Let the group of the device mount */
   { "nogroup",	0, 1, MS_GROUP  },	/* Device group has no special privs */
   { "_netdev",	0, 0, MS_COMMENT},	/* Device requires network */
+  { "_rnetdev", 0, 0, MS_COMMENT},	/* Device requires network, but should
+                                           be fscked by rc.sysinit */
   { "comment",	0, 0, MS_COMMENT},	/* fstab comment only (kudzu,_netdev)*/
 
   /* add new options here */
